java.lang.RuntimeException: java.lang.reflect.InvocationTargetException

Crash Report
============

This crash report was reported through the automatic crash reporting system 🤖

Traceback
--------------

```Python traceback
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:506)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)
Caused by: java.lang.reflect.InvocationTargetException
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
	... 1 more
Caused by: org.electroncash.electroncash3.ToastException: This wallet is watching-only.
	at org.electroncash.electroncash3.SendDialog.<init>(Send.kt:41)
	at org.electroncash.electroncash3.MainActivity.onOptionsItemSelected(Main.kt:255)
	at android.app.Activity.onMenuItemSelected(Activity.java:4182)
	at androidx.fragment.app.FragmentActivity.onMenuItemSelected(FragmentActivity.java:384)
	at androidx.appcompat.app.AppCompatActivity.onMenuItemSelected(AppCompatActivity.java:219)
	at androidx.appcompat.view.WindowCallbackWrapper.onMenuItemSelected(WindowCallbackWrapper.java:109)
	at androidx.appcompat.view.WindowCallbackWrapper.onMenuItemSelected(WindowCallbackWrapper.java:109)
	at androidx.appcompat.app.ToolbarActionBar$2.onMenuItemClick(ToolbarActionBar.java:64)
	at androidx.appcompat.widget.Toolbar$1.onMenuItemClick(Toolbar.java:207)
	at androidx.appcompat.widget.ActionMenuView$MenuBuilderCallback.onMenuItemSelected(ActionMenuView.java:781)
	at androidx.appcompat.view.menu.MenuBuilder.dispatchMenuItemSelected(MenuBuilder.java:840)
	at androidx.appcompat.view.menu.MenuItemImpl.invoke(MenuItemImpl.java:158)
	at androidx.appcompat.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:991)
	at androidx.appcompat.view.menu.MenuPopup.onItemClick(MenuPopup.java:128)
	at android.widget.AdapterView.performItemClick(AdapterView.java:374)
	at android.widget.AbsListView.performItemClick(AbsListView.java:1736)
	at android.widget.AbsListView$PerformClick.run(AbsListView.java:4207)
	at android.widget.AbsListView.onTouchUp(AbsListView.java:6707)
	at android.widget.AbsListView.onTouchEvent(AbsListView.java:6453)
	at android.widget.ListView.onTouchEvent(ListView.java:1721)
	at androidx.appcompat.widget.DropDownListView.onTouchEvent(DropDownListView.java:217)
	at androidx.appcompat.widget.MenuPopupWindow$MenuDropDownListView.onTouchEvent(MenuPopupWindow.java:135)
	at android.view.View.dispatchTouchEvent(View.java:14375)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3857)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:3535)
	at android.widget.AbsListView.dispatchTouchEvent(AbsListView.java:6365)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3863)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:3551)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3863)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:3551)
	at android.widget.PopupWindow$PopupDecorView.dispatchTouchEvent(PopupWindow.java:2860)
	at android.view.View.dispatchPointerEvent(View.java:14643)
	at android.view.ViewRootImpl$ViewPostImeInputStage.processPointerEvent(ViewRootImpl.java:6539)
	at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:6326)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5764)
	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5817)
	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5783)
	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:5939)
	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5791)
	at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:5996)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5764)
	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5817)
	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5783)
	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5791)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5764)
	at android.view.ViewRootImpl.deliverInputEvent(ViewRootImpl.java:8976)
	at android.view.ViewRootImpl.doProcessInputEvents(ViewRootImpl.java:8837)
	at android.view.ViewRootImpl.enqueueInputEvent(ViewRootImpl.java:8790)
	at android.view.ViewRootImpl$WindowInputEventReceiver.onInputEvent(ViewRootImpl.java:9112)
	at android.view.InputEventReceiver.dispatchInputEvent(InputEventReceiver.java:194)
	at android.os.MessageQueue.nativePollOnce(Native Method)
	at android.os.MessageQueue.next(MessageQueue.java:336)
	at android.os.Looper.loop(Looper.java:197)
	at android.app.ActivityThread.main(ActivityThread.java:8125)
	... 3 more
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
```

Reporter
------------

This issue was reported by 11 user(s):

| Electron Cash Version  | Python Version | Operating System  | Wallet Type  | Locale | Stack |
|---|---|---|---|---|---|
| 4.2.0-0  | 3.6.5 (default, Sep 11 2019, 15:39:45)  [GCC 4.9] | Android 10 on samsung SM-N960U1 (crownqlteue) | unknown | en_US | [ℹ](https://crashhub.electroncash.org//crash/1183) | 
| 4.2.0-0  | 3.6.5 (default, Sep 11 2019, 15:39:45)  [GCC 4.9] | Android 10 on samsung SM-G986W (y2q) | unknown | en_CA | [ℹ](https://crashhub.electroncash.org//crash/1184) | 
| 4.2.0-0  | 3.6.5 (default, Sep 11 2019, 15:30:12)  [GCC 4.9] | Android 10 on samsung SM-A115U (a11q) | unknown | en_US | [ℹ](https://crashhub.electroncash.org//crash/1195) | 
| 4.2.0-0  | 3.6.5 (default, Sep 11 2019, 15:30:12)  [GCC 4.9] | Android 10 on samsung SM-A115U (a11q) | unknown | en_US | [ℹ](https://crashhub.electroncash.org//crash/1228) | 
| 4.2.0-0  | 3.6.5 (default, Sep 11 2019, 15:39:45)  [GCC 4.9] | Android 10 on samsung SM-N960U1 (crownqlteue) | unknown | en_US | [ℹ](https://crashhub.electroncash.org//crash/1253) | 
| 4.2.0-0  | 3.6.5 (default, Sep 11 2019, 15:39:45)  [GCC 4.9] | Android 10 on samsung SM-A215U1 (a21) | unknown | en_US | [ℹ](https://crashhub.electroncash.org//crash/1259) | 
| 4.2.0-0  | 3.6.5 (default, Sep 11 2019, 15:39:45)  [GCC 4.9] | Android 10 on samsung SM-G960U (starqltesq) | unknown | en_US | [ℹ](https://crashhub.electroncash.org//crash/1283) | 
| 4.2.0-0  | 3.6.5 (default, Sep 11 2019, 15:39:45)  [GCC 4.9] | Android 10 on samsung SM-N975U (d2q) | unknown | en_US | [ℹ](https://crashhub.electroncash.org//crash/1300) | 
| 4.2.0-0  | 3.6.5 (default, Sep 11 2019, 15:39:45)  [GCC 4.9] | Android 10 on samsung SM-G970W (beyond0q) | unknown | en_CA | [ℹ](https://crashhub.electroncash.org//crash/1327) | 
| 4.2.0-0  | 3.6.5 (default, Sep 11 2019, 15:39:45)  [GCC 4.9] | Android 10 on samsung SM-G965U1 (star2qlteue) | unknown | vi_VN | [ℹ](https://crashhub.electroncash.org//crash/1386) | 
| 4.4.1-3  | 3.8.16 (default, Dec 25 2022, 19:06:35)  [Clang 11.0.5 (https://android.googlesource.com/toolchain/llvm-project 87f1315d | Android 10 on samsung SM-G965F (star2lte) | unknown | en_GB | [ℹ](https://crashhub.electroncash.org//crash/5267) | 


Additional Information
------------------------

The reporting user(s) did not provide additional information.

Good evening @mhsmith,

I just received another crash report related to this issue. The crash occured on Electron Cash 4.4.1-3.
I'm not sure which versions of Electron Cash include the fix but this is the first report from anything
newer than 4.2.0-0 since you closed the issue.

Could you please check if this issue really is resolved? Here is the traceback that I just collected:

```Python traceback
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:506)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)
Caused by: java.lang.reflect.InvocationTargetException
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
	... 1 more
Caused by: org.electroncash.electroncash3.ToastException: This wallet is watching-only.
	at org.electroncash.electroncash3.SendDialog.<init>(Send.kt:109)
	at org.electroncash.electroncash3.TokenTransactionsFragment.onViewCreated$lambda$6(TokenTransactions.kt:47)
	at org.electroncash.electroncash3.TokenTransactionsFragment.$r8$lambda$LTUupZDg5PqsZ5StldVi_K5lI9c(Unknown Source:0)
	at org.electroncash.electroncash3.TokenTransactionsFragment$$ExternalSyntheticLambda1.onClick(Unknown Source:2)
	at android.view.View.performClick(View.java:7862)
	at android.view.View.performClickInternal(View.java:7831)
	at android.view.View.access$3600(View.java:879)
	at android.view.View$PerformClick.run(View.java:29359)
	at android.os.Handler.handleCallback(Handler.java:883)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loop(Looper.java:237)
	at android.app.ActivityThread.main(ActivityThread.java:8167)
	... 3 more
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
```


~ _With robotic wishes_

--- a/android/app/src/main/java/org/electroncash/electroncash3/Requests.kt
+++ b/android/app/src/main/java/org/electroncash/electroncash3/Requests.kt
@@ -80,9 +80,6 @@ class NewRequestDialog : TaskDialog<PyObject>() {
     val listFragment by lazy { targetFragment as ListFragment }
 
     override fun doInBackground(): PyObject {
-        if (listFragment.wallet.callAttr("is_watching_only").toBoolean()) {
-            throw ToastException(R.string.this_wallet_is)
-        }
         return listFragment.wallet.callAttr("get_unused_address")
                ?: throw ToastException(R.string.no_more)
     }

--- a/android/app/src/main/java/org/electroncash/electroncash3/TokenTransactions.kt
+++ b/android/app/src/main/java/org/electroncash/electroncash3/TokenTransactions.kt
@@ -44,9 +44,13 @@ class TokenTransactionsFragment : ListFragment(R.layout.token_transactions, R.id
         binding.btnRequest.setOnClickListener { showDialog(this, NewRequestDialog().apply {
             arguments = Bundle().apply { putBoolean("token_request", true) }
         }) }
-        binding.btnSend.setOnClickListener { showDialog(this, SendDialog().apply {
-            arguments = Bundle().apply { putBoolean("token_send", true) }
-        }) }
+        binding.btnSend.setOnClickListener {
+            try {
+                showDialog(this, SendDialog().apply {
+                    arguments = Bundle().apply { putBoolean("token_send", true) }
+                })
+            } catch (e: ToastException) { e.show() }
+        }
     }
 
     override fun onCreateAdapter() = TokenTransactionsAdapter(this)

